---
layout: default
title: Displaying Annotation and Sequence Data from a GenBank Flat File
---

# Displaying Annotation and Sequence Data from a GenBank Flat File

This tutorial describes how to visualize the contents of a GenBank
flat file, including both the annotated sequence features and the DNA
sequence itself. Follow along step-by-step or use the following links
to jump to a specific section of interest:

1. [Install Circleator](#install_circleator)
2. [Download a GenBank flat file for a completed genome](#download_a_genbank_flat_file_for_a_completed_genome)
3. [Display the annotated genes](#display_the_annotated_genes)
4. [Add a percent GC-content plot](#add_a_percent_gccontent_plot)
5. [Highlight a specific gene or gene(s)](#highlighting_specific_genes)
6. [Download a GenBank flat file for an unfinished draft genome](#download_a_genbank_flat_file_for_an_unfinished_draft_genome)

### Install Circleator

To install Circleator please follow the instructions in the Circleator [Installation Guide][install].

[install]: /install.html

### Download a GenBank flat file for a completed genome

We'll start by drawing a Circleator figure for a single finished circular genome. 
For this example we'll use *Haemophilus influenzae Rd KW20*, the first microbe
ever to be sequenced. The GenBank flat file for its genome can be downloaded from
the NCBI web site as follows:

1. Go to <http://www.ncbi.nlm.nih.gov/nuccore/L42023>
2. Find the "Customize view" window in the top right corner of the page.
3. Select the "Show sequence" option under "Display options" and click on "Update View"
4. Wait for the sequence to be loaded (a notification bar should appear at the bottom of the page.)
5. Click on "Send" at the top right of the page and then select "File" under "Choose Destination"
6. Choose "GenBank (full)" for the Format and click on "Create File"
7. The GenBank entry should download into a file named "sequence.gb" (or possibly "sequence.gb (1)" 
or "sequence.gb (2)", etc., if this is not your first time doing this kind of thing.)
8. Check the first and last few lines of the file to make sure that it looks OK:

        $ head -5 ~/Downloads/sequence.gb 
        LOCUS       L42023               1830138 bp    DNA     circular BCT 23-OCT-2009
        DEFINITION  Haemophilus influenzae Rd KW20, complete genome.
        ACCESSION   L42023 U32686-U32848
        VERSION     L42023.1  GI:6626252
        DBLINK      BioProject: PRJNA219
        $ tail -5 ~/Downloads/sequence.gb 
          1830001 gatatagatc acaaaaaagt agtagggttt atagttttat aaaaatgctc gtgctatact
          1830061 ctgtgcgttg tcttactgag tgagcagtat tactcaaagc aaacagattt gtttaactta
          1830121 aataaaaggt gaaaatct
        //

9. Rename the downloaded file so we can remember what's in it:

        mv ~/Downloads/sequence.gb ./L42023.1.gb

### Display the annotated genes

Let's start by drawing just two things: an outer circle labeled with
sequence coordinates (0.0Mb, 0.5Mb, etc.) that represents the 1.83 Mb
*H. influenzae Rd KW20* genome, and an inner circle with a curved
black rectangle for each **gene** feature in the input file. The following
3-line Circleator configuration file ([genes-only.txt][]) accomplishes this goal:

    coords
    small-cgap
    genes

Note that each of these 3 things is a predefined track type. A complete list
of these can be found on the [predefined track types page][predef_tracks]. Now 
that we have both an input annotation file and a Circleator configuration file, 
all that remains is to run Circleator, like so:

    circleator --data=L42023.1.gb --config=genes-only.txt --pad=100 > hinf-genes-only.svg
    rasterize-svg hinf-genes-only.svg png 3000 3000

[predef_tracks]: /predefined-tracks.html

If the Circleator output includes a warning about "Unrecognized DBSOURCE data" 
this may be ignored: it's a warning generated by certain versions of BioPerl but
it should not affect the results. Here's what the Circleator output might look
like on the terminal after running the command above:

    INFO - started drawing figure using genes-only.txt 
    INFO - reading from annot_file=./L42023.1.gb, seq_file=, with seqlen=
    --------------------- WARNING --------------------- 
    MSG: Unrecognized DBSOURCE data: BioProject: PRJNA219
    --------------------------------------------------- 
    INFO - L42023: 3521 feature(s) and 1830138 bp of sequence 
    INFO - read 1 contig(s) from 1 input annotation and/or sequence file(s)
    INFO - finished drawing figure using genes-only.txt

And here is the resulting figure:

<div class='sample_image'>

<em>hinf-genes-only.png</em><br>
(config: <a href='gb_annotation/genes-only.txt'>genes-only.txt</a>, full size <a href='gb_annotation/hinf-genes-only-3000.png'>PNG</a>&nbsp;|&nbsp;<a href='gb_annotation/hinf-genes-only.svg'>SVG</a>)  
<img src='gb_annotation/hinf-genes-only-400.png' class='sample_image'>

</div>

[genes-only.txt]: gb_annotation/genes-only.txt

### Add a percent GC-content plot

Now let's display a simple quantity computed directly from the DNA
sequence, namely the percent GC-content. Percent GC-content is
typically plotted using a sliding window and by default Circleator
uses nonoverlapping windows. Here is our sample configuration file
([genes-and-percentGC-1.txt][]) with another small circular gap
(**small-cgap**) and a default percent GC plot (**%GC0-100**):

    coords
    small-cgap
    genes
    small-cgap
    %GC0-100

Run Circleator as before, passing it both the configuration file and
the GenBank flat file and directing the output into a .svg file, and
then rasterize the SVG file:

    circleator --data=L42023.1.gb --config=genes-and-percentGC-1.txt --pad=100 > hinf-genes-pctgc-1.svg
    rasterize-svg hinf-genes-pctgc-1.svg png 3000 3000

[genes-and-percentGC-1.txt]: gb_annotation/genes-and-percentGC-1.txt

Here is the resulting figure:

<div class='sample_image'>

<em>hinf-genes-pctgc-1.png</em><br>
(config: <a href='gb_annotation/genes-and-percentGC-1.txt'>genes-and-percentGC-1.txt</a>, full size <a href='gb_annotation/hinf-genes-pctgc-1-3000.png'>PNG</a>&nbsp;|&nbsp;<a href='gb_annotation/hinf-genes-pctgc-1.svg'>SVG</a>)  
<img src='gb_annotation/hinf-genes-pctgc-1-400.png' class='sample_image'>

</div>

The percent GC plot doesn't appear to be very informative, although if
you look at one of the full-size figures you can see that the percent
GC content stays right around the average value (38.1%) pretty much
throughout the entire genome.  In fact, if we run Circleator with more
verbose debugging enabled (`--debug=misc` or `--debug=all`) then it 
will tell us the observed minimum and maximum values for each of the 
graphs it draws, like so:

       DEBUG - graph data observed min=28.18, max=47.72, avg=38.129371717411, nvals=367, g_baseline=range_min, g_min=0, g_max=100

This tells us that although the **%GC0-100** track is drawn with range 0-100 (`g_min=0, g_max=100`)
the actual observed minimum value (for the given default window size) is 28.18 and 
the observed maximum value is 47.72. We can modify the percent-GC graph to
show more detail by decreasing the range accordingly, perhaps to 0-50
or even 25-50. We can also increase the height of the graph to show more
detail, as in the following example configuration file:

    coords
    small-cgap
    genes
    small-cgap
    # decrease graph_max from 100 to 50, increase graph_min to 25
    # and increase heightf from 0.07 to 0.15:
    %GC0-100 graph_max=50,graph_min=25,heightf=0.15
    # add a label to make it clear what's going on:
    medium-label label-text=Percent&nbsp;GC:&nbsp;25%-50%

Run Circleator and rasterize the resulting SVG file:

    circleator --data=L42023.1.gb --config=genes-and-percentGC-2.txt --pad=100 --debug=misc > hinf-genes-pctgc-2.svg
    rasterize-svg hinf-genes-pctgc-2.svg png 3000 3000

And the figure now looks like this:

<div class='sample_image'>

<em>hinf-genes-pctgc-2.png</em><br>
(config: <a href='gb_annotation/genes-and-percentGC-2.txt'>genes-and-percentGC-2.txt</a>, full size <a href='gb_annotation/hinf-genes-pctgc-2-3000.png'>PNG</a>&nbsp;|&nbsp;<a href='gb_annotation/hinf-genes-pctgc-2.svg'>SVG</a>)  
<img src='gb_annotation/hinf-genes-pctgc-2-400.png' class='sample_image'>

</div>

- gene-free region is only about 1kb (922039 - 922991)
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1160521/
find and highlight all low-gc regions, look for HI0687 HI0688 HI0051-HI0054